server/src/main/java/org/opensearch/search/pipeline/SystemGeneratedProcessorMetrics.java

```diff
@@ -0,0 +1,122 @@
+
+import org.opensearch.common.metrics.OperationMetrics;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+class SystemGeneratedProcessorMetrics {
+    private final Map<Processor.ProcessorType, Map<String, FactoryMetrics>> factoryMetrics = new HashMap<>();
+    private final Map<Processor.ProcessorType, Map<String, OperationMetrics>> processorMetrics = new HashMap<>();
+
+    SystemGeneratedProcessorMetrics() {
+        for (Processor.ProcessorType type : Processor.ProcessorType.values()) {
+            factoryMetrics.put(type, new HashMap<>());
+            processorMetrics.put(type, new HashMap<>());
+        }
+    }
+
+    void beforeEvaluation(String factoryType, Processor.ProcessorType processorType) {
+        getOrCreateFactoryMetrics(factoryType, processorType).evaluation.before();
+    }
+
+    void afterEvaluation(String factoryType, Processor.ProcessorType processorType, long timeInNanos) {
+        getOrCreateFactoryMetrics(factoryType, processorType).evaluation.after(timeInNanos);
+    }
+
+    void onEvaluationFailure(String factoryType, Processor.ProcessorType processorType) {
+        getOrCreateFactoryMetrics(factoryType, processorType).evaluation.failed();
+    }
+
+    void beforeGeneration(String factoryType, Processor.ProcessorType processorType) {
+        getOrCreateFactoryMetrics(factoryType, processorType).generation.before();
+    }
+
+    void afterGeneration(String factoryType, Processor.ProcessorType processorType, long timeInNanos) {
+        getOrCreateFactoryMetrics(factoryType, processorType).generation.after(timeInNanos);
+    }
+
+    void onGenerationFailure(String factoryType, Processor.ProcessorType processorType) {
+        getOrCreateFactoryMetrics(factoryType, processorType).generation.failed();
+    }
+
+    void beforeProcess(String processorKey, Processor.ProcessorType processorType) {
+        getOrCreateProcessorMetrics(processorKey, processorType).before();
+    }
+
+    void afterProcess(String processorKey, Processor.ProcessorType processorType, long timeInNanos) {
+        getOrCreateProcessorMetrics(processorKey, processorType).after(timeInNanos);
+    }
+
+    void onProcessFailure(String processorKey, Processor.ProcessorType processorType) {
+        getOrCreateProcessorMetrics(processorKey, processorType).failed();
+    }
+
+    private FactoryMetrics getOrCreateFactoryMetrics(String factoryType, Processor.ProcessorType processorType) {
+        return factoryMetrics.computeIfAbsent(processorType, k -> new HashMap<>()).computeIfAbsent(factoryType, k -> new FactoryMetrics());
+    }
+
+    private OperationMetrics getOrCreateProcessorMetrics(String processorKey, Processor.ProcessorType processorType) {
+        return processorMetrics.computeIfAbsent(processorType, k -> new HashMap<>())
+            .computeIfAbsent(processorKey, k -> new OperationMetrics());
```

Comment:
Nested computeIfAbsent calls could cause issues of concurrency if thread safe HashMap is not used.